app/controllers/addresses_controller.rb

```diff
@@ -0,0 +1,17 @@
+# frozen_string_literal: true
+
+class AddressesController < ApplicationController
+  def update
+    service = Service.find(params[:service_id])
+    address = Address.find(params[:id])
+
+    if !service.addresses.include?(address)
+      service.addresses << address
+      service.save!
+      render status: :created
+      return
+    else
+      render status: :precondition_failed
```

Comment:
Thinking about this more, we probably don't want to return a `precondition_failed` in the case where the relationship already exists. This is especially the case if we're going to use the PUT method for this API call, since you generally want [PUT requests to be idempotent](https://www.infoq.com/news/2013/04/idempotent/).

([Idempotence](https://en.wikipedia.org/wiki/Idempotence#Computer_science_meaning) is a fancy word for "If I call this function multiple times with the same arguments, it should have the same effect as calling it only once," or in practice, "If I accidentally call this a bunch of times with the same argument, it won't do anything different than if I just called it once like I had intended to.")

I think if the relationship exists already, then we can return successfully, though we should just return a 200 (OK) or a 204 (No Content) instead of a 201 (Created), per the [specification for PUT](https://www.w3.org/Protocols/rfc2616/rfc2616-sec9.html#sec9.6).